rdb/src/bestpath.rs

```diff
@@ -0,0 +1,194 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::BTreeSet;
+
+use crate::{db::Rib, types::Path, Prefix};
+use itertools::Itertools;
+
+/// The bestpath algorithms chooses the best set of up to `max` paths for a
+/// particular prefix from the RIB. The set of paths chosen will all have
+/// equal MED, local_pref, AS path lenght and shutdown status. The bestpath
+/// algorithm performs path filtering in the following ordered sequece of
+/// operations.
+///
+/// - partition candidate paths into active and shutdown groups.
+/// - if only shutdown routes exist, select from that group, otherwise
+///   select from the active group.
+/// - filter the selection group to the set of paths with the smallest
+///   local preference
+/// - filter the selection group to the set of paths with the smallest
+///   AS path lenght
+/// - filter the selection group to the set of paths with the smallest
+///   multi-exit discriminator (MED) on a per-AS basis.
+///
+/// Upon completion of these filtering operations, if the selection group
+/// is larger than `max`, return the first `max` entries. This is a set,
+/// so "first" has no semantic meaning, consider it to be random. If the
+/// selection group is smaller than `max`, the entire group is returned.
+pub fn bestpaths(
+    prefix: Prefix,
+    rib: &Rib,
+    max: usize,
+) -> Option<BTreeSet<Path>> {
+    let candidates = match rib.get(&prefix) {
+        Some(cs) => cs,
+        None => return None,
+    };
+
+    // Partition the choice space on whether routes are shutdown or not. If we
+    // only have shutdown routes then use those. Otherwise use active routes
+    let (active, shutdown): (BTreeSet<&Path>, BTreeSet<&Path>) =
+        candidates.iter().partition(|x| x.shutdown);
+    let candidates = match (active.len(), shutdown.len()) {
+        (0, _) => shutdown,
+        (_, _) => active,
+    };
+
+    // Filter down to paths that are not stale
```

Comment:
It wasn't very obvious to me how this worked at first, but maybe others are more
familiar with `itertools`. If you think this comment is valuable feel free to add it :)

```suggestion
    // Filter down to paths that are not stale
    // The `min_set_by_key` method allows us to assign "not stale" paths to the `0`
    // set, and "stale" paths to the `1` set. The method will then return the `0` set.
```

Reply:
Added